### ISSUE TITLE
CUDA device view is indexed by reference to host static variable

### ISSUE THREAD

**Issue by Char-Aznable** [kokkos/kokkos#1899]
Related to #1733. On CUDA backend, this causes silent memory error and it's almost impossible to debug. CUDA guide explicitly says that taking reference to host static variable causes error (see [here](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#const-variables)) but I haven't seen any warning or error from Clang when compiling code like this even though it corrupts the memory anyway:
```c++
class Foo {
   static constexpr int N = 10;
   ...
     //inside some device function Foo::f()
     view(N) = ...
   ...
};
```

**Comment by PhilMiller** [comment]
The versioned documentation link in question is  https://docs.nvidia.com/cuda/archive/9.2/cuda-c-programming-guide/index.html#const-variables

I find this a bit surprising, since there's also https://docs.nvidia.com/cuda/archive/9.2/cuda-c-programming-guide/index.html#constexpr-variables

Is the issue that `N` is being passed by reference to `View::operator()`?

**Comment by PhilMiller** [comment]
if the pass-by-reference is the issue, this may have actually gotten fixed pretty recently by #4800
